src/centrality/truss.cpp

```diff
@@ -0,0 +1,271 @@
+/*
+
+  Copyright 2017 The Johns Hopkins University Applied Physics Laboratory LLC. All Rights Reserved.
+  Copyright 2021 The igraph team.
+
+  Truss algorithm for cohesive subgroups.
+
+  Author: Alex Perrone
+  Date: 2017-08-03
+  Minor edits: The igraph team, 2021
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
+  02110-1301 USA
+
+*/
+
+#include "igraph_error.h"
+#include "igraph_adjlist.h"
+#include "igraph_interface.h"
+#include "igraph_motifs.h"
+#include "igraph_community.h"
+
+#include <iostream>
+#include <vector>
+#include <unordered_set>
+
+using namespace std;
+
+/* helper functions */
+void igraph_truss_i_unpack(const igraph_vector_int_t *tri, igraph_vector_t *unpacked_tri);
+void igraph_truss_i_compute_support(const igraph_vector_t *eid, igraph_vector_int_t *support);
+int igraph_i_trussness(const igraph_t *graph, igraph_vector_int_t *support,
+  igraph_vector_int_t *truss);
+
+/**
+ * \function trussness
```

Comment:
The name here must match the function.

Wouldn't `igraph_trussness` be a better name for this function? Opinions?

Reply:
I had the same thought. The truss is the subgraph, but because of inclusivity of course one can reconstruct the subgraphs from this function.

I suggest we add two functions to the API:
- `igraph_trussness` -> the current `igraph_truss`, but with the last argument renames to `trussness`
- `int igraph_truss(const igraph_t *graph, long int k, igraph_t *truss)`

And also let's use the convention of everyone else without subtracting -2 from the trussness

Reply:
Hmm, maybe this idea is not so good after all. If we want to give the user the actual subgraph, we need to initialize a new graph, but then vertex ids will not match unless we include a bunch of singletons (then it's not the actualy truss, technically). For k-cores, we only have one function `igraph_coreness`, let's do the same here.